### PR TITLE
Laboratory work #2, Denis Zaytsev - 22FPL1

### PR DESCRIPTION
**Pull request by shoodeen**

### CODE REVIEW

File: lab_2_tokenize_by_bpe/main.py [hunk 1]
```diff
@@ -26,6 +35,16 @@ def collect_frequencies(
...
     """
+    if not isinstance(text, str) or not isinstance(start_of_word, str | None) \
+            or not isinstance(end_of_word, str):
+        return None
+    word_frequencies = {}
+    for word in text.split():
+        processed_word = prepare_word(word, start_of_word, end_of_word)
+        if processed_word is None:
+            return None
+        word_frequencies.update({processed_word: text.split().count(word)})
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:46]: Consider checking if the word is already in the dictionary first.

File: lab_2_tokenize_by_bpe/main.py [hunk 2]
```diff
@@ -36,6 +55,18 @@ def count_tokens_pairs(
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+    token_pairs_frequency = {}
+    for word, frequency in word_frequencies.items():
+        if len(word) >= 2:
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:62]: Consider doing an early return here.

File: lab_2_tokenize_by_bpe/main.py [hunk 3]
```diff
@@ -36,6 +55,18 @@ def count_tokens_pairs(
...
+            for i in range(len(word) - 1):
+                token_pair = (word[i], word[i + 1])
+                if token_pair in token_pairs_frequency:
+                    token_pairs_frequency[token_pair] += frequency
+                else:
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:67]: Can we avoid using `else` here?

File: seminars/practice_2_string.py
Comment (marina-kaz) [inline seminars/practice_2_string.py:1]: Please, remove changes in the files other that in the folder with lab2

**Review by marina-kaz — COMMENTED** [review]
Good start! Keep going!

File: lab_2_tokenize_by_bpe/main.py [hunk 4]
```diff
@@ -14,6 +14,15 @@ def prepare_word(
     :param end_of_word: a token that signifies the end of word
     :return: preprocessed word
     """
+    if not isinstance(raw_word, str) or not isinstance(start_of_word, str | None)\
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:17]: `isinstance` is not an optimal way to check if value is `None`: as it is a singleton, consider using `is` instead

**Review by marina-kaz — COMMENTED** [review]
Not much changed since my last review

**Review by marina-kaz — COMMENTED** [review]
No changes since my last review

**Review by marina-kaz — COMMENTED** [review]
A kind reminder that the deadline is this week! We only have one round of review left, consider speeding up!

**Review by marina-kaz — COMMENTED** [review]
No changes still

**Comment by shoodeen** [comment]
I don't know why, but the changes I pushed yesterday didn't appear here

**Review by marina-kaz — COMMENTED** [review]
Still no changes

**Review by marina-kaz — COMMENTED** [review]
Nothing to review here today

**Pull request closed without merging**